/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.proyectounidad2.persistencia;

import com.mycompany.proyectounidad2.dominio.Hobby;
import jakarta.persistence.EntityManager;
import jakarta.persistence.TypedQuery;
import java.util.List;

/**
 *
 * @author Afgord
 */
public class HobbyDAO implements IHobbyDAO {

    private final EntityManager em;

    public HobbyDAO(EntityManager em) {
        this.em = em;
    }

    @Override
    public Hobby guardar(Hobby hobby) {
        em.persist(hobby);
        return hobby;
    }

    @Override
    public Hobby actualizar(Hobby hobby) {
        return em.merge(hobby);
    }

    @Override
    public Hobby buscarPorId(Long id) {
        return em.find(Hobby.class, id);
    }

    @Override
    public Hobby buscarPorNombre(String nombre) {
        String jpql = """
                SELECT h
                FROM Hobby h
                WHERE h.nombre = :nombre
                """;

        TypedQuery<Hobby> query = em.createQuery(jpql, Hobby.class);
        query.setParameter("nombre", nombre);

        return query.getResultStream().findFirst().orElse(null);
    }

    @Override
    public List<Hobby> obtenerTodos() {
        String jpql = "SELECT h FROM Hobby h";
        return em.createQuery(jpql, Hobby.class).getResultList();
    }

}
